File: lib/connection.js
var _ = require('lodash'),
  rest = require('restler'),
  Promise = require('bluebird'),
  Retry = require('./retry'),
  errors = require('./errors'),
  util = require('util'),
  sha512 = require('js-sha512'),
  EventEmitter = require('events').EventEmitter,
  config = require('../config/config'),
  log = require('./util').logger(),
  restlerMethodArgCount = {
    get: 2,
    post: 2,
    put: 2,
    del: 2,
    head: 2,
    patch: 2,
    json: 3,
    postJson: 3,
    putJson: 3
  };

function Connection(options) {
  EventEmitter.call(this);
  this._options = options || {};
  this._tokenData = this._options.accessToken ? { access_token: this._options.accessToken } : null;
  this._retry = new Retry(this._options.retry || {});
  this.apiCallCount = 0;
  _.each(restlerMethodArgCount, function (count, method) {
    this[method] = function (var_args) {
      var args = Array.prototype.slice.call(arguments);
      while (args.length < count) {
        args.push(undefined);
      }
      args.unshift(method);
      return this._request.apply(this, args);
    };
  }.bind(this));
}
util.inherits(Connection, EventEmitter);

function getNextPageFn(conn, method, args) {
  var options = _.clone(_.last(args) || {});
  args = _.clone(args);
  args.pop();
  return function nextPage(nextPageToken) {
    var params = options;
    if (method === 'get') params = options.query = options.query || {};
    else if (method === 'post' || method === 'put') params = options.data = options.data || {};
    params.nextPageToken = nextPageToken;
    return conn._request.apply(conn, _.flatten([method, args, options], true));
  }
}

Connection.prototype._request = function (method) {
  var args = Array.prototype.slice.call(arguments, 1),
    url = _.first(args);
  if (!/^https?:\/\//i.test(url)) {
    var baseUrl = this._options.url || config.api.url;
    args[0] = baseUrl + url;
  }
  this.emit('apiCall', { apiCallCount: ++this.apiCallCount });
  log.debug('Request:', arguments);
  var requestFn = function requestFn(forceOAuth) {
    var iframeCtx = this._options.iframeCtx;
    return this.getOAuthToken(forceOAuth, iframeCtx).then(function (token) {
      var defer = Promise.defer(),
        options = _.last(args) || {},
        nextPageFn = getNextPageFn(this, method, args);
      args.pop();
      options.headers = _.extend({}, options.headers, {
        Authorization: 'Bearer ' + token.access_token
      });
      args.push(options);
      rest[method].apply(rest, args)
        .on('success', function (data, resp) {
          if (data.success === false && _.has(data, 'errors')) {
            log.debug('Request failed: ', data);
            defer.reject(data);
          } else {
            if (_.has(data, 'nextPageToken')) {
              Object.defineProperty(data, 'nextPage', {
                enumerable: false,
                value: _.partial(nextPageFn, data.nextPageToken)
              });
            }
            defer.resolve(data);
          }
        })
        .on('error', function (err, resp) {
          defer.reject(err);
        })
        .on('fail', function (data, resp) {
          if (_.has(data, 'errors')) {
            log.debug('Request failed: ', data);
            return defer.reject(data);
          }
          var statusCode = Math.floor(resp.statusCode / 100);
          if (statusCode === 5) defer.reject(new errors.Http5XXError(data));
          else if (statusCode === 4) defer.reject(new errors.Http4XXError(data));
          else defer.reject(data);
        })
        .on('timeout', function (ms) {
          defer.reject(new errors.TimeoutError(ms));
        });
      return defer.promise;
    }.bind(this), function (e) {
      throw new Error('Authentication (' + e.error + '): ' + e.error_description);
    });
  };
  return this._retry.start(requestFn, this);
};

Connection.prototype.getIframeAuthToken = function (iframeCtx, url) {
  var secret = this._options.secret || '';
  return sha512(iframeCtx.uid.toString() + iframeCtx.ts.toString() + (url || '') + secret);
}

Connection.prototype.getOAuthToken = function (force, iframeCtx) {
  force = force || false;
  if (force || this._tokenData == null) {
    var defer = Promise.defer();
    var requestFn = function () {
      var options = {
        data: {
          grant_type: this.grantType || (iframeCtx ? 'client_credentials' : 'password'),
          client_id: this._options.clientId,
          client_secret: this._options.clientSecret,
          username: this._options.username,
          password: this._options.password,
          // scope: this._options.scope || 'oob',
        },
        timeout: this._options.timeout || 20000
      };
      var baseUrl = this._options.url || config.api.url;
      rest.post(baseUrl + '/auth/oauth/v2/token', options)
        .on('success', function (data, resp) {
          if (iframeCtx) {
            log.debug('Got pre-token: ', data);
            this.getOnBehalfOAuthToken(data, iframeCtx).then(function (token) {
              this._tokenData = token;
              defer.resolve(token);
            }, function () {
              defer.reject();
            });
          }
          else {
            log.debug('Got token: ', data);
            this._tokenData = data;
            defer.resolve(data);
          }
        }.bind(this))
        .on('timeout', function (ms) {
          log.debug('Timeout');
          defer.reject(ms);
        })
        .on('fail', function (data, resp) {
          defer.reject(data);
        })
        .on('error', function (err) {
          defer.reject(err);
        });
      return defer.promise;
    };
    return this._retry.start(requestFn, this);
  } else {
    log.debug('Using existing token: ', this._tokenData);
    return Promise.resolve(this._tokenData);
  }
};

Connection.prototype.getOnBehalfOAuthToken = function (tokenData, iframeCtx) {
  var defer = Promise.defer();
  var requestFn = function () {
    var getOptions = {
      data: { memberId: iframeCtx.uid },
      headers: { Authorization: 'Bearer ' + tokenData },
      timeout: this._options.timeout || 20000
    };
    var baseUrl = this._options.url || config.api.url;
    rest.post(baseUrl + '/v1/tokens', options)
      .on('success', function (data, resp) {
        log.debug('Got token: ', data);
        this._tokenData = data;
        defer.resolve(data);
      }.bind(this))
      .on('timeout', function (ms) {
        log.debug('Timeout');
        defer.reject(ms);
      })
      .on('fail', function (data, resp) {
        defer.reject(data);
      })
      .on('error', function (err) {
        defer.reject(err);
      });
    return defer.promise;
  };
  return this._retry.start(requestFn, this);
};

module.exports = Connection;

File: src.net/Hootsuite.Rest/Connection.cs
using Hootsuite.Rest.Require;
using Newtonsoft.Json.Linq;
using System;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

//https://stackoverflow.com/questions/38996593/promise-equivalent-in-c-sharp
//https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/console-webapiclient
//https://dotnetcodr.com/2014/03/25/continuation-tasks-in-net-tpl-exception-handling-in-task-chains/
namespace Hootsuite.Rest
{
    public class Connection
    {
        readonly dynamic _options;
        readonly Retry _retry;
        readonly Restler _rest = new Restler();
        readonly Action<string> _log = util.logger;
        JObject _tokenData;

        public class FrameContext
        {
            public string lang { get; set; }
            public string timezone { get; set; }
            public string pid { get; set; }
            public string uid { get; set; }
            public string ts { get; set; }
            public string token { get; set; }
        }

        public Connection(dynamic options)
        {
            _options = options ?? new { };
            _tokenData = dyn.hasProp(_options, "accessToken") ? dyn.ToJObject(new { access_token = _options.accessToken }) : null;
            _retry = new Retry(dyn.getProp(_options, "retry", new { }));
        }

        public Task<JObject> get(string url, dynamic options = null) => _request(url, null, options, Restler.Method.GET);
        public Task<JObject> post(string url, dynamic options = null) => _request(url, null, options, Restler.Method.POST);
        public Task<JObject> put(string url, dynamic options = null) => _request(url, null, options, Restler.Method.PUT);
        public Task<JObject> del(string url, dynamic options = null) => _request(url, null, options, Restler.Method.DELETE);
        public Task<JObject> head(string url, dynamic options = null) => _request(url, null, options, Restler.Method.HEAD);
        public Task<JObject> patch(string url, dynamic options = null) => _request(url, null, options, Restler.Method.PATCH);
        public Task<JObject> json(string url, object data, dynamic options = null) => _request(url, data, options, Restler.Method.GET);
        public Task<JObject> postJson(string url, object data, dynamic options = null) => _request(url, data, options, Restler.Method.POST);
        public Task<JObject> putJson(string url, object data, dynamic options = null) => _request(url, data, options, Restler.Method.PUT);

        static readonly Regex HttpTestExp = new Regex(@"^https?://", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.Singleline);

        private async Task<JObject> _request(string url, object data, dynamic options, Restler.Method method)
        {
            options = dyn.exp(options, true);
            if (!HttpTestExp.IsMatch(url))
            {
                var baseUrl = dyn.getProp(_options, "url", config.api_url);
                url = baseUrl + url;
            }
            _log($"Request: {url}");
            Func<Task<JObject>, Task<JObject>> requestFn0 = async (x) =>
            {
                if (!x.IsFaulted)
                {
                    var token = x.Result;
                    options.headers = dyn.getObj(options, "headers");
                    options.headers.Authorization = $"Bearer {token["access_token"]}";
                    try
                    {
                        var y = await _rest.request(url, options, method);
                        var d = (JObject)y;
                        if ((string)d["success"] == "false" && d["errors"] != null)
                        {
                            _log($"Request failed: {d}");
                            throw new InvalidOperationException(d["errors"].ToString());
                        }
                        return d;
                    }
                    catch (RestlerOperationException res)
                    {
                        var err = (JObject)res.Content;
                        if (err["errors"] != null)
                        {
                            _log($"Request failed: {err}");
                            throw;
                        }
                        else
                        {
                            var statusCode = Math.Floor((int)res.StatusCode / 100M);
                            if (statusCode == 5) throw res;
                            else if (statusCode == 4) throw res;
                            else throw res;
                        }
                    }
                }
                else
                {
                    //    //var firstError = "Error"; // e.errors[0];
                    //    //throw new InvalidOperationException($"Authentication ({firstError.code}): {firstError.message}");
                    var res = x.Exception;
                    _log("ERROR");
                    throw res;
                }
            };
            Func<bool, Task<JObject>> requestFn = async (forceOAuth) => await GetOAuthToken(forceOAuth).ContinueWith(x =>
            {
                return requestFn0(x).Result;
            });
            return await _retry.start(requestFn, this);
        }

        private string GetFrameAuthToken(string url)
        {
            var secret = dyn.getProp<string>(_options, "secret");
            var frameCtx = dyn.getProp<dynamic>(_options, "frameCtx");
            using (var hmac = new HMACSHA512(secret))
            {
                var bytes = Encoding.UTF8.GetBytes(frameCtx.uid + frameCtx.ts + (url ?? string.Empty));
                var hash = hmac.ComputeHash(bytes);
                return Convert.ToBase64String(hash);
            }
        }

        private async Task<JObject> GetOAuthToken(bool force = false)
        {
            if (force || _tokenData == null)
            {
                var frameCtx = dyn.getProp<dynamic>(_options, "frameCtx");
                Func<bool, Task<JObject>> requestFn = async (forceOAuth) =>
                {
                    var options = new
                    {
                        data = new
                        {
                            grant_type = dyn.getProp(_options, "grantType", (frameCtx != null ? "client_credentials" : "password")),
                            client_id = dyn.getProp(_options, "clientId", (string)null),
                            client_secret = dyn.getProp(_options, "clientSecret", (string)null),
                            username = dyn.getProp(_options, "username", (string)null),
                            password = dyn.getProp(_options, "password", (string)null),
                            // scope= dyn.getProp(_options, "scope", "oob"),
                        },
                        timeout = dyn.getProp(_options, "timeout", 20000),
                    };
                    var baseUrl = dyn.getProp(_options, "url", config.api_url);
                    try
                    {
                        var x = await _rest.post($"{baseUrl}/auth/oauth/v2/token", options);
                        var data = (JObject)x;
                        if (frameCtx != null)
                        {
                            _log($"Got pre-token: {data}");
                            try { return await GetOnBehalfAuthToken(data); }
                            catch (Exception err) { _log("ERROR"); throw err; }
                        }
                        else
                        {
                            _log($"Got token: {data}");
                            _tokenData = data;
                            return data;
                        }
                    }
                    catch (Exception err) { _log("ERROR"); throw err; }
                };
                return await _retry.start(requestFn, this);
            }
            else
            {
                _log($"Using existing token: {_tokenData}");
                return _tokenData;
            }
        }

        private async Task<JObject> GetOnBehalfAuthToken(dynamic tokenData)
        {
            var frameCtx = dyn.getProp(_options, "frameCtx", new { uid = string.Empty });
            Func<bool, Task<JObject>> requestFn = async (forceOAuth) =>
            {
                var options = new
                {
                    data = new { memberId = frameCtx.uid },
                    headers = new { Authorization = $"Bearer {tokenData.access_token}" },
                    timeout = dyn.getProp(_options, "timeout", 20000),
                };
                var baseUrl = dyn.getProp(_options, "url", config.api_url);
                try
                {
                    var x = await _rest.postJson($"{baseUrl}/v1/tokens", options.data, options);
                    var data = (JObject)x;
                    _log($"Got token: {data}");
                    _tokenData = data;
                    return data;
                }
                catch (Exception err) { _log("ERROR"); throw err; }
            };
            return await _retry.start(requestFn, this);
        }
    }
}
